src/settings/controllers/SlidingSyncController.ts

```diff
@@ -14,18 +15,72 @@ See the License for the specific language governing permissions and
 import { SettingLevel } from "../SettingLevel";
-import { SlidingSyncOptionsDialog } from "../../components/views/dialogs/SlidingSyncOptionsDialog";
-import Modal from "../../Modal";
 import SettingsStore from "../SettingsStore";
 import { _t } from "../../languageHandler";
+import { MatrixClientPeg } from "../../MatrixClientPeg";
 
 export default class SlidingSyncController extends SettingController {
+    /**
+     * Check if the server declares support for sliding sync via a proxy in its client `.well-known`.
+     * @return {boolean} Whether the client well-known contains a proxy url
+     */
+    private async proxySlidingSyncSupport(): Promise<boolean> {
+        const clientWellKnown = MatrixClientPeg.safeGet().getClientWellKnown();
+        const proxyUrl = clientWellKnown?.["org.matrix.msc3575.proxy"]?.url;
+        return proxyUrl != undefined;
+    }
+
+    /**
+     * Check if the server declares "natively" supports sliding sync (at the unstable endpoint)
+     * @return {boolean} Whether the "native" (unstable) endpoint is up
+     */
+    private async nativeSlidingSyncSupport(): Promise<boolean> {
+        const cli = MatrixClientPeg.safeGet();
+        return cli.http.authedRequest(Method.Post, "/sync", undefined, undefined, {
+            localTimeoutMs: 10 * 1000, // 10s
+            prefix: "/_matrix/client/unstable/org.matrix.msc3575",
+        });
+    }
+
+    /**
+     * Check whether our homeserver has sliding sync support, that the endpoint is up, and
+     * is a sliding sync endpoint.
+     * @return {boolean} Whether the server has a working sliding sync implementation
+     */
+    private async slidingSyncHealthCheck(): Promise<boolean> {
+        let baseUrl: String
+        if (await this.proxySlidingSyncSupport()) {
+            baseUrl = MatrixClientPeg.safeGet().getClientWellKnown()?.["org.matrix.msc3575.proxy"]!.url;
+        }
+        else if (await this.nativeSlidingSyncSupport()) {
+            baseUrl = MatrixClientPeg.safeGet().getHomeserverUrl();
+        }
+        else {
+            return false;
+        }
+
+        const controller = new AbortController();
+        const id = window.setTimeout(() => controller.abort(), 10 * 1000); // 10s
+        const res = await fetch(baseUrl + "/client/server.json", {
+            signal: controller.signal,
+        });
+        clearTimeout(id);
+        if (res.status === 200) {
+            logger.info("slidingSyncHealthCheck: sliding sync endpoint is up");
+            return true;
+        }
+
+        return false;
+    }
+
     public async beforeChange(level: SettingLevel, roomId: string, newValue: any): Promise<boolean> {
-        const { finished } = Modal.createDialog(SlidingSyncOptionsDialog);
-        const [value] = await finished;
+        const value = await this.slidingSyncHealthCheck();
         return newValue === value; // abort the operation if we're already in the state the user chose via modal
     }
```

Comment:
Not sure we need this as the setting will be disabled if unsupported and would just yield a confusing UX of the toggle not working with no user feedback as to why